bundles/binding/org.openhab.binding.smlreader/src/main/java/org/openhab/binding/smlreader/internal/SmlDevice.java

```diff
@@ -0,0 +1,314 @@
+	*
+	* @param byte to convert to Integer.
+	*/
+	private int byteToInt(byte b){
+		return Integer.parseInt(String.format("%02x", b), 16);
+	}
+
+	/**
+	* Decodes native SML informations from the device and stores them locally until the next read request.
+	*
+	* @param smlFile the native SML informations from the device
+	*/
+	private void populateValueCache(SML_File smlFile) {
+		do {
+			if(smlFile == null) {
+				logger.warn("{}: no valid SML File.", this.toString());
+				break;
+			}
+
+			List<SML_Message> smlMessages = smlFile.getMessages();
+
+			if(smlMessages == null) {
+				logger.warn("{}: no valid SML messages list retrieved.", this.toString());
+				break;
+			}
+			
+			int messageCount = smlMessages.size();
+			
+			if(messageCount <= 0) {
+				logger.warn("{}: SML message list does not have entries.", this.toString());
+				break;
+			}
+			
+			for (int i = 0; i < messageCount; i++) {
+				SML_Message sml_message = smlMessages.get(i);
+				
+				if(sml_message == null) {
+					logger.warn("{}: no valid SML message.", this.toString());
+					continue;
+				}				
+
+				int tag = sml_message.getMessageBody().getTag().getVal();
+				
+				if(tag != SML_MessageBody.GetListResponse) {
+					continue;
+				}
+				
+				SML_GetListRes listResponse = (SML_GetListRes)sml_message.getMessageBody().getChoice();
+				SML_List smlValueList = listResponse.getValList();
+				SML_ListEntry[] smlListEntries = smlValueList.getValListEntry();
+
+				for (SML_ListEntry entry : smlListEntries) {
+					String obis = getObisAsString(entry.getObjName().getOctetString());
+					
+					SmlValue smlValue = valueCache.get(obis);
+					
+					if(smlValue == null){
+						smlValue = new SmlValue(entry);	
+					}
+
+					valueCache.put(obis, smlValue);
+				}
+			}
+			
+			printInfo();
+			
+		}while(false);
+	}
+	
+	/**
+	 * Gets the configured deviceId.
+	 *
+	 * @return the id of the SmlDevice from openHAB configuration.
+	 */
+	public String getDeviceId()	{
+		return deviceId;
+	}
+	
+	/**
+	 * Returns the specified OBIS value if available.
+	 *
+	 * @param obis the OBIS code which value should be retrieved.
+	 * @return the OBIS value as String if available - otherwise null.
+	 */
+	public String getValue(String obisId){
+		String obisValue = null;
+		
+		if(valueCache.containsKey(obisId)){
+			obisValue = valueCache.get(obisId).getValue();
+		}
+			
+		return obisValue;
+	}
+
+	/**
+	* Read values from this device an store them locally against their OBIS code. 
+	*/
+	public void readValues() {
+		SML_File smlFile = null;
+		
+		do {
+
+			clearValueCache();
+			
+			if(connector == null){
+				logger.error(this.toString() + " connector is not instantiated");
+				break;
+			}
+			
+			try {
+				smlFile = connector.getMeterValues();
+			} catch(ConnectorException ex) {
+				logger.error(this.toString() + ": Error during receive values from device: " + ex.getMessage());
+				break;
+			}
+						
+			populateValueCache(smlFile);
+		
+		}while (false);
```

Comment:
It's just a way to avoid the amount of nesting the code to execute. I think it will give the focus on code than on the correct brace/tab (especially in more complex functions). So if it's not quite clear i'm going to workaround the method - no problem.